### PR TITLE
REGR: fix consolidation/cache issue with take operation

### LINKED ISSUE #35521: REGR: invalid cache after take operation with non-consolidated dataframe
a program run ok with pandas 1.0.5,  but after upgrade to 1.1.0.  it fail.
i found sometimes when I change one cell of dataframe value,  print(df) and it doesn't change.
but i use  ==, it says changed.
for example:  original  cell value is A, then i change cell value to B.   A=B
print df, it show still A
df.at(x, x) == B, it says TRUE.

-----

MarcoGorelli's edit: here's a reproducible example:

```
import pandas as pd


position = pd.DataFrame(columns=["code", "startdate"])
position = position.append([{"code": "a", "startdate": 0}])

# These two lines should not change anything.
# BUT, commenting either of them out makes this code run as intendeed
position["code"] == "A"
position[position["startdate"] == 0]

position.at[0, "code"] = "A"

print(position.at[0, "code"])
print(position)
```

output:
```
A
  code startdate
0    a         0
```

expected output:
```
A
  code startdate
0    A         0
```

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Closes #35521

@jbrockmendel the "short term" fix for 1.1.2

### CODE REVIEW

File: doc/source/whatsnew/v1.1.2.rst [hunk 1]
```diff
@@ -17,6 +17,7 @@ Fixed regressions
 - Regression in :meth:`DatetimeIndex.intersection` incorrectly raising ``AssertionError`` when intersecting against a list (:issue:`35876`)
 - Fix regression in updating a column inplace (e.g. using ``df['col'].fillna(.., inplace=True)``) (:issue:`35731`)
 - Performance regression for :meth:`RangeIndex.format` (:issue:`35712`)
+- Fix regression in invalid cache after an indexing operation (:issue:`35521`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.2.rst:20]: can you elaborate a bit here, this is pretty general

File: pandas/tests/frame/test_block_internals.py [hunk 2]
```diff
@@ -658,3 +658,26 @@ def test_update_inplace_sets_valid_block_values():
 
     # smoketest for OP bug from GH#35731
     assert df.isnull().sum().sum() == 0
+
+
+def test_nonconsolidated_item_cache_take():
```
Comment (jreback) [inline pandas/tests/frame/test_block_internals.py:663]: i would put in pandas/tests/frame/test_indexing.py (or similar) rather than here.

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm minor comments

File: doc/source/whatsnew/v1.1.2.rst (hunk 1 above)
Comment (jorisvandenbossche) [reply doc/source/whatsnew/v1.1.2.rst:20]: If you have a suggestion, happy to change it. But it's a rather obscure issue that is hard to explain in understandable terms

File: pandas/tests/frame/test_block_internals.py (hunk 2 above)
Comment (jorisvandenbossche) [reply pandas/tests/frame/test_block_internals.py:663]: It's not really an indexing bug, but rather a bug in `take` in combination with indexing before and afterwards (and it's the assign operation that only makes the bug visible), and is specific to the consolidation state of the original dataframe. So it's more a bug in the internal machinery of consolidation / cache.

File: doc/source/whatsnew/v1.1.2.rst (hunk 1 above)
Comment (jorisvandenbossche) [reply doc/source/whatsnew/v1.1.2.rst:20]: OK, that's nice, thanks!

**Comment by jbrockmendel** [comment]
The non-copying alternative would be to not call `self._consolidate_inplace` in `BlockManager.take`

**Comment by jbrockmendel** [comment]
LGTM; can look into the no-copy solution longer-term

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Comment by jorisvandenbossche** [comment]
@MeeseeksDev backport to 1.1.x

**Pull request merged**